syntax = "proto3";

package conversations;

import "google/protobuf/timestamp.proto";
import "google/protobuf/wrappers.proto";
import "google/protobuf/empty.proto";

service Conversations {
  rpc ListMessageThreads(ListMessageThreadsReq) returns (ListMessageThreadsRes) {
    // Gives a paginated list of message threads with previews
  }

  rpc EditMessageThreadStatus(EditMessageThreadStatusReq) returns (google.protobuf.Empty);

  rpc GetMessageThread(GetMessageThreadReq) returns (GetMessageThreadRes);

  rpc GetMessageThreadInfo(GetMessageThreadInfoReq) returns (GetMessageThreadInfoRes);

  rpc CreateMessageThread(CreateMessageThreadReq) returns (CreateMessageThreadRes);

  rpc EditMessageThread(EditMessageThreadReq) returns (google.protobuf.Empty);

  rpc MakeMessageThreadAdmin(ThreadUserReq) returns (google.protobuf.Empty);

  rpc RemoveMessageThreadAdmin(ThreadUserReq) returns (google.protobuf.Empty);

  rpc SendMessage(SendMessageReq) returns (google.protobuf.Empty);

  rpc LeaveMessageThread(LeaveMessageThreadReq) returns (google.protobuf.Empty);

  rpc InviteToMessageThread(ThreadUserReq) returns (google.protobuf.Empty);

  rpc SearchMessages(SearchMessagesReq) returns (SearchMessagesRes);
}

enum MessageThreadStatus {
  PENDING = 0;
  ACCEPTED = 1;
  REJECTED = 2;
}

message MessageThreadPreview {
  int64 thread_id = 1;
  string title = 2;
  repeated string recipients = 3;
  bool is_dm = 4;
  google.protobuf.Timestamp creation_time = 5;
  MessageThreadStatus status = 6;
  string latest_message_preview = 7;
  string latest_message_sender = 8;
  google.protobuf.Timestamp latest_message_time = 9;
}

message ListMessageThreadsReq {
  uint32 start_index = 1;
  uint32 max = 2;
}

message ListMessageThreadsRes {
  uint32 start_index = 1;
  repeated MessageThreadPreview threads = 2;
  bool has_more = 3;
}


message EditMessageThreadStatusReq {
  int64 thread_id = 1;
  MessageThreadStatus status = 2;
}

message Message {
  int64 id = 1;
  string sender = 2;
  google.protobuf.Timestamp timestamp = 3;
  string text = 4;
}

message GetMessageThreadReq {
  int64 thread_id = 1;
  uint32 start_index = 2;
  uint32 max = 3;
}

message GetMessageThreadRes {
  uint32 start_index = 1;
  repeated Message messages = 2;
  bool has_more = 3;
}

message GetMessageThreadInfoReq {
  uint32 thread_id = 1;
}

message GetMessageThreadInfoRes {
  string title = 1;
  repeated string recipients = 2;
  repeated string admins = 3;
  google.protobuf.Timestamp creation_time = 4;
  bool only_admins_invite = 5;
  MessageThreadStatus status = 6;
  bool is_dm = 7;
}

message CreateMessageThreadReq {
  google.protobuf.StringValue title = 1;
  repeated string recipients = 2;
}

message CreateMessageThreadRes {
  int64 thread_id = 1;
}

message EditMessageThreadReq {
  int64 thread_id = 1;
  google.protobuf.StringValue title = 2;
  google.protobuf.BoolValue only_admins_invite = 3;
}

message ThreadUserReq {
  int64 thread_id = 1;
  string user = 2;
}

message SendMessageReq {
  int64 thread_id = 1;
  string message = 2;
}

message LeaveMessageThreadReq {
  int64 thread_id = 1;
}

message SearchMessagesReq {
  string query = 1;
}

message MessageSearchResult {
  int64 thread_id = 1;
  int64 message_id = 2;
}

message SearchMessagesRes {
  repeated MessageSearchResult results = 1;
}
